src/Microsoft.AspNetCore.Server.Kestrel/BadHttpResponseException.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Http;
+
+namespace Microsoft.AspNetCore.Server.Kestrel
+{
+    public sealed class BadHttpResponseException : InvalidOperationException
```

Comment:
I wouldn't consider these server created exceptions. These are application created exceptions. Most APIs will throw plain `InvalidOperationException`s if called incorrectly.

I actually don't mind `BadHttpResponseException` for the caller attempting to do things that would break the HTTP spec like setting `Transfer-Encoding` and a non-body status. I personally wouldn't do it, but I don't mind.

Throwing a `BadHttpResponseException` for calling `OnStarted` too late or attempting to add response headers too late seems confusing and wrong though.